### DISCUSSION TITLE
Quality End-of-Speech Detection while Live Streaming in busy retail environments

### DISCUSSION THREAD

**Discussion by lincolnsmiff** [deepgram/community#882] [General help] [answered]
I'm working on a project involving live streaming audio transcription in busy retail environments, and I need some clarity on implementing accurate end-of-speech detection.

Currently, we use interim results to display near real-time transcription and the `UtteranceEnd` event to finalize and update the transcription shortly after an `is_final=true` message.

Here are the specific parameters we are using in our WebSocket connection:
```
wss://api.deepgram.com/v1/listen?smart_format=true&filler_words=true&speech_final=true&interim_results=true&utterance_end_ms=1000&endpoint=300&is_final=true
```

We want to understand how to effectively use `endpointing` in addition to `UtteranceEnd`. We believe that utterances are based on inferred gaps or pauses between words, while endpointing is based on actual audible sound gaps or silence.

### Our Current Understanding:
1. Trigger 'update to final message and send' when a transcript with `speech_final=true` is received.
   - This may be followed by an `UtteranceEnd` message, which can be ignored in this case.
2. Trigger 'update to final message and send' if an `UtteranceEnd` message is received without a preceding `speech_final=true` message.

### Questions:
1. Does the above logic therefore assume that `endpointing` is more effective than the `UtteranceEnd` message, and is this simply because this transcription audio has been processed faster than its inferred words? Or what other purpose would it serve in our scenario?
2. Given that WebSocket returns the transcription with the `speech_final=true` flag, do we use this transcription directly, and is there no additional message like there is with `UtteranceEnd`?
3. Is our goal in the real world to configure the audio quality and parameters (such as `utterance_end_ms` delay and `endpointing` delay) to ensure `is_final=true` and `speech_final=true` occur simultaneously (given this is processed in 1 second chunks)? (In this case, we assume we would use the transcription and ignore the following `UtteranceEnd` message).
4. If we receive an `UtteranceEnd` message, is there any scenario where we should expect a subsequent `speech_final=true` message?

Much appreciated!

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by jkroll-deepgram** [comment, marked as answer]
Hi @lincolnsmiff, all great questions and important considerations for using Deepgram in a streaming application!

We have quite a few doc pages on these topics, and you've likely read at least some of them, but I'll share the full list for you to consult.
[Endpointing & Interim Results With Live Streaming](https://developers.deepgram.com/docs/understand-endpointing-interim-results)
[End of Speech Detection While Live Streaming](https://developers.deepgram.com/docs/understanding-end-of-speech-detection)
[Interim Results](https://developers.deepgram.com/docs/interim-results)
[Endpointing](https://developers.deepgram.com/docs/endpointing)
[Utterance End](https://developers.deepgram.com/docs/utterance-end)

And to address each of your questions:
1. As you mentioned, endpointing uses a VAD (i.e. noise/audio itself), while Utterance End uses the transcript. Using both sources of information helps you be robust to background noise in your audio in cases where endpointing isn't as responsive.
2. Yes, you can use the transcription directly that includes the `speech_final=true` flag.
3. No, typically you'll want either/or logic to be able to infer a possible end-of-speech from either `speech_final` or `UtterancEnd`. Also note that `speech_final` indicates a likely pause in speech, while `is_final` will occur every 3-5 seconds, indicating that we have highest confidence in our transcript of the last several seconds (in contrast to interim results, which we return every 1 second and are a preliminary transcript).
4. Yes, since `UtteranceEnd` indicates that no word has been transcribed, and `speech_final=true` indicates directly a pause in speech, one could follow the other. However, `speech_final` is typically in the range of 10s to 100s of milliseconds, while `UtteranceEnd` has a minimum of 1 second, so typically `speech_final` would trigger before `UtteranceEnd`, not after.

**Reply by lincolnsmiff** [reply]
Thanks for the reply. I had reviewed those articles. As feedback, it would be great if your tutorials included more detailed implementation examples that combine multiple features and offer fine-tuning tips. We are currently fine-tuning the `utterance_end_ms` and `endpointing` parameters.

Based on your feedback, here’s our understanding and approach to avoid broken messages:

1. Increase the `endpointing` delay from the default until we stop getting broken phrases. This will be tested without background noise. Use the transcription directly with the `speech_final=true` flag.

2. If background noise prevents `endpointing` from triggering `speech_final`, we will use the `UtteranceEnd` message, based on gaps in word transcription. To test we would turn off `endpointing`, then adjust `utterance_end_ms` to the lowest value that avoids broken messages, likely higher than the 1000ms default since that level caused issues for us.

Note: We noticed that both `speech_final=true` and `UtteranceEnd` break messages in two at their default levels, but better to be able to go up than not be able to go down...

Is this approach correct for minimizing broken phrases and handling background noise effectively? Any further insights or recommendations for fine-tuning these settings would be greatly appreciated.

Thank you for your support!

**Reply by jkroll-deepgram** [reply]
Hi @lincolnsmiff, that all sounds like a good path forward.

I want to share once more resource. We grappled with this type of voicebot and end-of-speech detection for our own [browser-based AI Agent demo](https://deepgram-conversational-demo.vercel.app/). We've open-sourced it; here is the [top-level repo](https://github.com/deepgram-devs/deepgram-ai-agent-demo), along with the [Deepgram speech-to-text settings](https://github.com/deepgram-devs/deepgram-ai-agent-demo/blob/ddbcdd8c03f4310e641b4de11b1f45e34085520e/app/context/Deepgram.tsx#L44-L51) and the section [handling streaming events](https://github.com/deepgram-devs/deepgram-ai-agent-demo/blob/ddbcdd8c03f4310e641b4de11b1f45e34085520e/app/components/Conversation.tsx#L196-L315).

One more note: it won't be of use to you yet, but we understand that end-of-thought detection is crucial for AI agents, and we're in the research phase of training models that actually incorporate all available audio and conversational context in order to directly output a prediction as to whether a user has finished their thought and stopped speaking. Stay tuned for some exciting announcements and feature releases in the coming months.

**Reply by lincolnsmiff** [reply]
Thanks Julia, that repo should be useful to my dev team. I'll get them stuck into it.

Wow, yes, end-of-thought would be super useful. I'm sure you probably don't need the input, but I can't help myself; training your models on voice inflections and not just the words like 'wellll...' or 'uhh' would be amazingly useful for streaming (perhaps even to extend the utterance end or endpointing delay at those points). And I'm sure it'd be useful in dictation: even if the thought doesn't get finished or gets cut off, you could end the phrase with a–

**Reply by jkroll-deepgram** [reply]
@lincolnsmiff, I appreciate the feedback and insight into how you would envision the feature being used! I've saved your feedback into our product tracker for end-of-thought detection, for reference by our product team.
